In [47]:
import pandas as pd
import numpy as np

data_file="Inclusive_Exclusive race sentences - Sheet1.csv"

exclusive_column = "Exclusive"
inclusive_column = "Inclusive"

data = pd.read_csv(data_file)
data_exc = pd.DataFrame(data[exclusive_column])
data_inc = pd.DataFrame(data[inclusive_column])

data_exc_labeled = data_exc
data_exc_labeled['inc'] = '0'
data_exc_labeled.rename(columns={exclusive_column:'text'}, inplace = True)

data_inc_labeled = data_inc
data_inc_labeled['inc'] = '1'
data_inc_labeled.rename(columns={inclusive_column:'text'}, inplace = True)

data_both = pd.concat([data_exc_labeled, data_inc_labeled])

print(data_both.shape)
print(data_both.head())

(54, 2)
                                                text inc
0  African immigrants are only able to seek unqua...   0
1         Roma/Romani people live in extreme poverty   0
2  The Commissioner welcomed a group of Asian vis...   0
3  Roma/Romani people had a meeting with the Comm...   0
4    I met a group of African American people today.   0


In [48]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=1000)
vectors = vectorizer.fit_transform(data_both.text)
words_df = pd.DataFrame(vectors.toarray(), columns=vectorizer.get_feature_names_out())
words_df.head()

able  achieve  affect  africa   african  all  always  american  \
0  0.388024      0.0     0.0     0.0  0.234697  0.0     0.0  0.000000   
1  0.000000      0.0     0.0     0.0  0.000000  0.0     0.0  0.000000   
2  0.000000      0.0     0.0     0.0  0.000000  0.0     0.0  0.000000   
3  0.000000      0.0     0.0     0.0  0.000000  0.0     0.0  0.000000   
4  0.000000      0.0     0.0     0.0  0.315558  0.0     0.0  0.410906   

   americans  and  ...  welcomed  welfare  what  where  white      with  \
0        0.0  0.0  ...  0.000000      0.0   0.0    0.0    0.0  0.000000   
1        0.0  0.0  ...  0.000000      0.0   0.0    0.0    0.0  0.000000   
2        0.0  0.0  ...  0.439151      0.0   0.0    0.0    0.0  0.000000   
3        0.0  0.0  ...  0.000000      0.0   0.0    0.0    0.0  0.385843   
4        0.0  0.0  ...  0.000000      0.0   0.0    0.0    0.0  0.000000   

   working  worse  you  your  
0      0.0    0.0  0.0   0.0  
1      0.0    0.0  0.0   0.0  
2      0.0    0.0  0.0   0.0  
3      0.0    0.0  0.0   0.0  
4      0.0    0.0  0.0   0.0  

[5 rows x 188 columns]

In [49]:
X = words_df
y = data_both.inc

In [50]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB

now we are going to use all of the above algorithms^ to create a basline on this dataset to make an argument for DNN.

In [51]:
#Logistic Regression.
%%time
logreg = LogisticRegression(C=1e9, solver='lbfgs', max_iter=1000)
logreg.fit(X, y)

CPU times: user 11.3 ms, sys: 11.7 ms, total: 22.9 ms
Wall time: 19.4 ms


LogisticRegression(C=1000000000.0, max_iter=1000)

In [52]:
#Random Forest
%%time
forest = RandomForestClassifier(n_estimators=50)
forest.fit(X, y)

CPU times: user 92.7 ms, sys: 62.8 ms, total: 156 ms
Wall time: 96 ms


RandomForestClassifier(n_estimators=50)

In [53]:
# LinearSVC
%%time
svc = LinearSVC()
svc.fit(X, y)

CPU times: user 4.26 ms, sys: 923 µs, total: 5.19 ms
Wall time: 4.65 ms


LinearSVC()

In [54]:
# MultinomialNB
%%time
bayes = MultinomialNB()
bayes.fit(X, y)

CPU times: user 5.07 ms, sys: 0 ns, total: 5.07 ms
Wall time: 4.48 ms


MultinomialNB()

Using our models!!  
This is a simple test set I am using.

In [55]:
data_file="SimpleClassifierTestSet - Sheet1.csv"
data_test = pd.read_csv(data_file)
print(data_test.shape)
print(data_test)

(8, 2)
                                                text  inc
0  A white person would be well suited for this job.    0
1                 Anyone that is smart will do well.    1
2                       Asians do not speak English.    0
3         Everyone has the potential to be a doctor.    1
4              All people deserve equal opportunity.    1
5                   Africans Americans are criminals    0
6          Immigrants should have less voting power.    0
7                  Do not judge a book by its cover.    1


In [56]:
print(vectorizer.get_feature_names_out())

['able' 'achieve' 'affect' 'africa' 'african' 'all' 'always' 'american'
 'americans' 'and' 'are' 'as' 'asia' 'asian' 'asians' 'aspects' 'bad' 'be'
 'because' 'beer' 'being' 'believe' 'benefit' 'best' 'biased' 'black'
 'bomb' 'built' 'can' 'cannot' 'color' 'commissioner' 'common'
 'communites' 'congo' 'criminals' 'culture' 'curry' 'did' 'difficult'
 'disciplined' 'disparities' 'do' 'does' 'dominant' 'due' 'eastern'
 'easterns' 'eat' 'edjucation' 'enjoy' 'equal' 'ethnically' 'ethnicity'
 'everything' 'existing' 'experience' 'extreme' 'fair' 'family' 'for'
 'fork' 'from' 'future' 'garden' 'german' 'germans' 'group' 'grow' 'had'
 'hands' 'hard' 'has' 'have' 'he' 'here' 'heritage' 'higher' 'historical'
 'illegaly' 'immigrants' 'immigrate' 'in' 'incriminate' 'indians'
 'indigenous' 'inherently' 'is' 'it' 'japan' 'jobs' 'judicial' 'kenya'
 'korea' 'lack' 'lead' 'live' 'looks' 'love' 'many' 'meeting' 'met'
 'middle' 'migrants' 'more' 'must' 'native' 'nigeria' 'not' 'of' 'often'
 'on' 'only' 'o

In [65]:
# Put it through the vectoriser

# transform, not fit_transform, because we already learned all our words
data_test_vectors = vectorizer.transform(data_test.text)
data_test_words_df = pd.DataFrame(data_test_vectors.toarray(), columns=vectorizer.get_feature_names_out())
print(data_test_words_df.head())
print(data_test_words_df.shape)

   able  achieve  affect  africa  african       all  always  american  \
0   0.0      0.0     0.0     0.0      0.0  0.000000     0.0       0.0   
1   0.0      0.0     0.0     0.0      0.0  0.000000     0.0       0.0   
2   0.0      0.0     0.0     0.0      0.0  0.000000     0.0       0.0   
3   0.0      0.0     0.0     0.0      0.0  0.000000     0.0       0.0   
4   0.0      0.0     0.0     0.0      0.0  0.571508     0.0       0.0   

   americans  and  ...  welcomed  welfare  what  where     white  with  \
0        0.0  0.0  ...       0.0      0.0   0.0    0.0  0.465444   0.0   
1        0.0  0.0  ...       0.0      0.0   0.0    0.0  0.000000   0.0   
2        0.0  0.0  ...       0.0      0.0   0.0    0.0  0.000000   0.0   
3        0.0  0.0  ...       0.0      0.0   0.0    0.0  0.000000   0.0   
4        0.0  0.0  ...       0.0      0.0   0.0    0.0  0.000000   0.0   

   working  worse  you  your  
0      0.0    0.0  0.0   0.0  
1      0.0    0.0  0.0   0.0  
2      0.0    0.0  0.0 

Our data is all stored in Tf/Idf...  
so there is 188 columns, one for each word used in all training and test data.

In [58]:
# Predict using all our models. 

# Logistic Regression predictions + probabilities
data_test['pred_logreg'] = logreg.predict(data_test_words_df)
data_test['pred_logreg_proba'] = logreg.predict_proba(data_test_words_df)[:,1]

# Random forest predictions + probabilities
data_test['pred_forest'] = forest.predict(data_test_words_df)
data_test['pred_forest_proba'] = forest.predict_proba(data_test_words_df)[:,1]

# SVC predictions
data_test['pred_svc'] = svc.predict(data_test_words_df)

# Bayes predictions + probabilities
data_test['pred_bayes'] = bayes.predict(data_test_words_df)
data_test['pred_bayes_proba'] = bayes.predict_proba(data_test_words_df)[:,1]

In [59]:
data_test

text  inc pred_logreg  \
0  A white person would be well suited for this job.    0           1   
1                 Anyone that is smart will do well.    1           1   
2                       Asians do not speak English.    0           0   
3         Everyone has the potential to be a doctor.    1           1   
4              All people deserve equal opportunity.    1           1   
5                   Africans Americans are criminals    0           0   
6          Immigrants should have less voting power.    0           0   
7                  Do not judge a book by its cover.    1           0   

   pred_logreg_proba pred_forest  pred_forest_proba pred_svc pred_bayes  \
0       9.956683e-01           0               0.40        1          1   
1       5.123794e-01           1               0.60        1          0   
2       3.711442e-12           0               0.28        0          0   
3       9.999717e-01           1               0.58        1          1   
4       9.853761e-01           0               0.36        1          1   
5       1.369807e-14           0               0.18        0          0   
6       7.596668e-07           0               0.36        0          0   
7       8.909374e-11           0               0.32        0          0   

   pred_bayes_proba  
0          0.550030  
1          0.438844  
2          0.287532  
3          0.533914  
4          0.516234  
5          0.225517  
6          0.383369  
7          0.331688

### Time for a confusion matrix for each of the results!!  
this is much easier to understand results this way.

In [60]:
from sklearn.metrics import confusion_matrix

y_test = data_test["inc"].astype(str)
X_test = data_test_words_df

In [61]:
y_pred_logreg = logreg.predict(X_test)
matrix = confusion_matrix(y_test, y_pred_logreg)

label_names = pd.Series(['negative', 'positive'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted negative  Predicted positive
Is negative                   3                   1
Is positive                   1                   3

In [62]:
y_pred_forest = forest.predict(data_test_words_df)
matrix = confusion_matrix(y_test, y_pred_forest)

label_names = pd.Series(['negative', 'positive'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted negative  Predicted positive
Is negative                   4                   0
Is positive                   2                   2

In [63]:
y_pred_svc = svc.predict(data_test_words_df)
matrix = confusion_matrix(y_test, y_pred_svc)

label_names = pd.Series(['negative', 'positive'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted negative  Predicted positive
Is negative                   3                   1
Is positive                   1                   3

In [64]:
y_pred_bayes = bayes.predict(data_test_words_df)
matrix = confusion_matrix(y_test, y_pred_bayes)

label_names = pd.Series(['negative', 'positive'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted negative  Predicted positive
Is negative                   3                   1
Is positive                   2                   2